### PR TITLE
Refs #160: Sync flow and result object format optimizations.

### PR DESCRIPTION
**Pull request by n1k0**
Following discussion on #160 ([this comment](https://github.com/Kinto/kinto.js/issues/160#issuecomment-140712336)), I've experimented with removing the second call to `pullChanges`, and to my great surprise all the tests were mostly passing.

I went further, and now locally touched records after that second pull are no more listed in the `updated` array property of the sync result object — that was clearly misleading.

I've also found a bug which was preventing encoded records to be decoded when updated after having been published; also, the `published` sync result property now lists records in their decoded flavor — which is probably more user-friendly.

I'm having hard times figuring out if I'm missing any obvious use case with this patch, but I'm still pretty confident its viable when it comes to all functional tests being green.

Also admire the concision of that new version of the`sync` method; no mention we're gaining one HTTP request, which is good from a performances standpoint.

Would love hearing thoughts from @michielbdejong @Natim @ametaireau @leplatrem and anybody interested in this topic.

**Update**

Had to reintroduce that second call to `pullChanges` in order to address possible race conditions in-between synchronization steps. Other addressed points mentioned above remain.

### CODE REVIEW

**Comment by n1k0** [comment]
Gosh, it's picky.

screen shot 2015-09-16 at 16 52 27

File: src/collection.js [hunk 1]
```diff
@@ -506,7 +506,8 @@ export default class Collection {
             });
           } else {
             // Remote update was successful, reflect it locally
-            return this.update(record, {synced: true});
+            return this._decodeRecord("remote", record)
```
Comment (Natim) [inline src/collection.js]: Nice catch.

**Comment by leplatrem** [comment]
> anybody interested in this topic.

I'm sorry I cannot dig into this deeply. But it looks good to me. AFAIU one drawback is that you will download your own pushed records during the next sync (because `this._lastModified` is only updated during `importChanges()`).

Side note: don't forget to update the diagram :)

**Comment by n1k0** [comment]
> (because `this._lastModified` is only updated during `importChanges()`).

Yeah, that was basically the main purpose of that second pull, as well as force-marking remote version as synced locally.

> Side note: don't forget to update the diagram :)

Good point!

**Comment by n1k0** [comment]
After playing with the demo using that patch, I'm finding annoying to have to sync twice in order to have an all clear sync result object. I still need to work out how to properly mark published records as synced locally, with the proper `last_modified` value updated with the right data (which are obtained in the batch request individual responses I think).

**Comment by n1k0** [comment]
Well, it appears the only way I have to update the remote collection `lastModified` value from the server is indeed by pulling changes from it; and we absolutely need to update this value to ensure next sync operation will use the latest known timestamp… Would it be conceivable to have the latest collection `lastModified` value retrieved from a BATCH main response?

File: docs/api.md [hunk 2]
```diff
@@ -291,24 +291,40 @@ articles.sync({
...
+  deleted:   [],
+  skipped:   [],
+  published: [],
+  resolved:  [],
 }
 ```
 
+The synchronization result object exposes the following properties:
+
+- `ok`: The boolean status of the synchronization operation; `true` if no unsolvable conflicts and no errors were encountered.
```
Comment (michielbdejong) [inline docs/api.md]: unsolvable -> unresolved - presumably, conflicts that are handled manually are in themselves solvable, it just hasn't happened yet. :)

File: docs/api.md [hunk 3]
```diff
@@ -291,24 +291,40 @@ articles.sync({
...
+- `lastModified`: The timestamp of the latest revision of server data.
```
Comment (michielbdejong) [inline docs/api.md]: Is that one timestamp for the whole collection? This timestamp is in milliseconds since 1970, right? Does the server guarantee that any changes that happen immediately after one of the clients synced happen in the next millisecond?

File: docs/api.md [hunk 4]
```diff
@@ -291,24 +291,40 @@ articles.sync({
...
+- `conflicts`: The list of encountred conflicts during both import and export operations (see *[Resolving conflicts manually](#resolving-conflicts-manually)*);
```
Comment (michielbdejong) [inline docs/api.md]: But only the ones that were encountered and that require manual resolution, right?

Nit: encountred -> encountered

File: docs/api.md [hunk 5]
```diff
@@ -291,24 +291,40 @@ articles.sync({
...
+- `errors`:    The list of encountered errors, if any.
```
Comment (michielbdejong) [inline docs/api.md]: Per record (so one 401 error at the http-level can result in 20 errors at the record-level here). Is it guaranteed that a record that is listed here was left unchanged? For instance, if a record is pulled in, but the RemoteTransformer throws an error, is the lastModified value reset to (or left at) what it was, so that it will be retried next time?

File: docs/api.md [hunk 6]
```diff
@@ -291,24 +291,40 @@ articles.sync({
...
+- `created`:   The list of remote records which have been succesfully imported into the local database.
+- `updated`:   The list of remote record updates which have been succesfully reflected into the local database.
+- `deleted`:   The list of remotely deleted records which have been succesfully deleted as well locally.
+- `skipped`:   The list of identical records which have been declared as synced automatically.
+- `published`: The list of locally modified records which have been succesfully pushed to the remote server.
```
Comment (michielbdejong) [inline docs/api.md]: "modified" here meaning "created, updated, or deleted". But I guess that's clear enough.

File: docs/api.md [hunk 7]
```diff
@@ -291,24 +291,40 @@ articles.sync({
...
+- `resolved`:  The list of conflicting records which have been succesfully resolved according to the selected strategy (note that using the default `MANUAL` strategy, this list is always empty).
```
Comment (michielbdejong) [inline docs/api.md]: Nit: succesfully -> successfully

File: src/collection.js [hunk 8]
```diff
@@ -506,7 +506,8 @@ export default class Collection {
             });
           } else {
             // Remote update was successful, reflect it locally
```
Comment (michielbdejong) [inline src/collection.js]: Nit: Remote create or update

**Comment by michielbdejong** [comment]
> Would it be conceivable to have the latest collection lastModified value retrieved from a BATCH main response?

That wouldn't help you. Say you pulled at 12:00, and then publish a new record at 12:07, and learn that the lastModified of the collection as a whole now equals the `response.data.last_modified` from [the response body of the POST](http://kinto.readthedocs.org/en/latest/api/records.html#records-post) that you just did. That still tells you nothing, another client may have made some other change at 12:03.

I think the best option is to leave the "last full sync" timestamp for the collection at 12:00 and set `this.lastModified` of just the record that you published at 12:07, to 12:07 (using the `response.data.last_modified` value from that request response). Then, at 12:10 do another sync, in which you will see for instance:
* new collection timestamp: 12:07
* record created at 12:03 (this is a change that still needs to be reflected locally)
* record created at 12:07 (the one you published yourself, so ignore)

**Comment by michielbdejong** [comment]
It's just a pity that all data that you publish is also sent back to you again, which can be quite inefficient if records contain big blobs of data. You're basically guaranteed to have twice as much traffic as needed for outgoing changes.

If batch operations are atomic, then maybe it would make sense to do a GET at the beginning of the batch, and then thanks to batch atomicity you could safely set the collection timestamp to the highest one from the batch.

Or maybe add per-record _since timestamps to the GET query, but that would become quite an ugly API, probably.

File: docs/api.md (hunk 3 above)
Comment (Natim) [inline docs/api.md]: > Is that one timestamp for the whole collection?

Yes

> This timestamp is in milliseconds since 1970, right?

Yes

> Does the server guarantee that any changes that happen immediately after one of the clients synced happen in the next millisecond?

Yes.

File: docs/api.md (hunk 5 above)
Comment (Natim) [inline docs/api.md]: Changes are individual for the list so if 1 worked over 20, the collection timestamp changed and the record that worked changed. But the other records are left unchanged.

**Comment by Natim** [comment]
> It's just a pity that all data that you publish is also sent back to you again, which can be quite inefficient if records contain big blobs of data.

Yes this is a known limitation and we have a way to fix it. See #144

**Comment by n1k0** [comment]
I think I'll just revert the change removing that second call, which is the best way to ensure any race is properly addressed during the flow; #144 will address part of the performance concerns. We may come back to this later anyway.

File: docs/api.md (hunk 5 above)
Comment (michielbdejong) [inline docs/api.md]: I guess the collection timestamp always stays before the oldest record that errored, to make sure all errors get endless retry until they succeed?

Comment (n1k0) [inline docs/api.md]: Nope, and that's a good point, I'll investigate that.

Comment (n1k0) [inline docs/api.md]: Filed #163 about it.

**Comment by n1k0** [comment]
Failed build says:

```
$ nvm install 0.12

Version '0.12' not found - try `nvm ls-remote` to browse available versions.
```

Now that's new -_-

**Comment by n1k0** [comment]
This was a temporary issue on Travis it seems. This is now ready for review; r=? @michielbdejong @Natim

**Comment by Natim** [comment]
It looks good to me.

File: docs/api.md [hunk 9]
```diff
@@ -291,24 +291,40 @@ articles.sync({
...
+  skipped:   [],
+  published: [],
+  resolved:  [],
 }
 ```
 
+The synchronization result object exposes the following properties:
+
+- `ok`: The boolean status of the synchronization operation; `true` if no unresolved conflicts and no errors were encountered.
+- `lastModified`: The timestamp of the latest revision of server data.
```
Comment (michielbdejong) [inline docs/api.md]: This description can be expanded and made more precise as part of #163. For now it's close enough, I think.

File: docs/api.md [hunk 10]
```diff
@@ -291,24 +291,40 @@ articles.sync({
...
+- `conflicts`: The list of encountered conflicts during both import and export operations (see *[Resolving conflicts manually](#resolving-conflicts-manually)*);
```
Comment (michielbdejong) [inline docs/api.md]: Only the unresolved ones, not the resolved ones. But the reader will probably understand that.

File: test/integration_test.js [hunk 11]
```diff
@@ -167,10 +167,8 @@ describe("Integration tests", () => {
           expect(cleanRecord(syncResult.published[0])).eql(testData.localUnsynced[0]);
         });
 
-        it("should mark local records as synced", () => {
```
Comment (michielbdejong) [inline test/integration_test.js]: Are we still testing that it does actually mark the local record as synced? I guess we were testing that implicitly through the inclusion in`updated`, and now we're not testing it at all anymore? (at least not here).

File: docs/api.md (hunk 9 above)
Comment (michielbdejong) [inline docs/api.md]: An example of when this description is incorrect is if there are unresolved conflicts in the SyncResult for instance, see this test https://github.com/Kinto/kinto.js/blob/sync-result-updated-list/test/integration_test.js#L352

File: test/integration_test.js [hunk 12]
```diff
@@ -779,10 +1059,10 @@ describe("Integration tests", () => {
             ]);
           });
 
-          it("should encode records when pushed to the server", () => {
```
Comment (michielbdejong) [inline test/integration_test.js]: Are we still testing that it actually does do the encoding. The single biggest bug risk in terms of implications is if unencrypted data leaks out of a Kinto.js-based app due to some bug where the RemoteTransformer does not get applied in some edge case, so we should have supergood integration tests for that.

Comment (michielbdejong) [inline test/integration_test.js]: Ah, found it. https://github.com/Kinto/kinto.js/blob/sync-result-updated-list/test/integration_test.js#L1068

**Comment by michielbdejong** [comment]
r+ from me too. (maybe create a follow-up for https://github.com/Kinto/kinto.js/pull/161/files#r39848183 if needed).

Thanks for addressing this! It makes the SyncResults a lot clearer.

File: test/integration_test.js (hunk 11 above)
Comment (n1k0) [inline test/integration_test.js]: Yes, in the [very next test](https://github.com/Kinto/kinto.js/blob/sync-result-updated-list/test/integration_test.js#L174-L187).

File: test/integration_test.js (hunk 12 above)
Comment (n1k0) [inline test/integration_test.js]: Yeah, using transofrmers end to end integration tests, we're pretty much safe here.

File: docs/api.md (hunk 10 above)
Comment (n1k0) [inline docs/api.md]: It's worth a precision, will update.
